### PR TITLE
Fix sonarqube using base branch on fork for detecting new code in pr

### PR DESCRIPTION
**Pull request by t3chguy**
Attempt number 2 from https://github.com/matrix-org/matrix-js-sdk/pull/2378

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: .github/workflows/sonarcloud.yml [hunk 1]
```diff
@@ -57,13 +50,20 @@ jobs:
...
-        run: git rev-parse HEAD && git fetch origin ${{ inputs.main_branch }}:${{ inputs.main_branch }} && git status && git rev-parse HEAD
+      # Fetch base branch from the upstream repo so that Sonar can identify new code in PR builds
+      - name: "📕 Fetch upstream base branch"
+        # workflow_call retains the github context of the caller, so `action_repository` will be upstream always
+        # due to it running on `workflow_run` which is called from the context of the target repo and not the fork.
+        if: inputs.base_branch && github.action_repository != inputs.repo
+        run: |
+          git remote add upstream https://github.com/${{ github.action_repository }}.git
+          git rev-parse HEAD
+          git fetch origin ${{ inputs.base_branch }}:${{ inputs.base_branch }}
```
Comment (weeman1337) [inline .github/workflows/sonarcloud.yml:61]: Shouldn't that be `git fetch upstream`? If that is correct can you please tweak the comments?

**Review by weeman1337 — CHANGES_REQUESTED** [review]

File: .github/workflows/sonarcloud.yml (hunk 1 above)
Comment (t3chguy) [reply .github/workflows/sonarcloud.yml:61]: Good spot

File: .github/workflows/sonarcloud.yml [hunk 2]
```diff
@@ -57,13 +50,20 @@ jobs:
...
-      - name: "📕 Fetch ${{ inputs.main_branch }}"
-        if: inputs.head_branch != inputs.main_branch
-        run: git rev-parse HEAD && git fetch origin ${{ inputs.main_branch }}:${{ inputs.main_branch }} && git status && git rev-parse HEAD
+      # Fetch base branch from the upstream repo so that Sonar can identify new code in PR builds
+      - name: "📕 Fetch upstream base branch"
+        # workflow_call retains the github context of the caller, so `action_repository` will be upstream always
+        # due to it running on `workflow_run` which is called from the context of the target repo and not the fork.
+        if: inputs.base_branch && github.action_repository != inputs.repo
+        run: |
+          git remote add upstream https://github.com/${{ github.action_repository }}.git
```
Comment (t3chguy) [inline .github/workflows/sonarcloud.yml:59]: Turns out `action_repository` is not what I expected it to be, needs fixing

**Comment by t3chguy** [comment]
/me puts back on the general queue due to unavailability

File: .github/workflows/sonarcloud.yml [hunk 3]
```diff
@@ -57,13 +51,20 @@ jobs:
...
 
-      # Fetch develop so that Sonar can identify new issues in PR builds
-      - name: "📕 Fetch ${{ inputs.main_branch }}"
-        if: inputs.head_branch != inputs.main_branch
-        run: git rev-parse HEAD && git fetch origin ${{ inputs.main_branch }}:${{ inputs.main_branch }} && git status && git rev-parse HEAD
+      # Fetch base branch from the upstream repo so that Sonar can identify new code in PR builds
+      - name: "📕 Fetch upstream base branch"
+        # workflow_call retains the github context of the caller, so `repository` & `repositoryUrl` will be upstream
+        # always due to it running on `workflow_run` which is called from the context of the target repo and not the fork.
+        if: inputs.base_branch && github.repository != inputs.repo
```
Comment (robintown) [inline .github/workflows/sonarcloud.yml:58]: If this part of the job only runs on forked PRs, what part of this checks out the base branch on non-forked PRs?

**Review by robintown — APPROVED** [review]

**Pull request merged**